src/scalarstats.jl

```diff
@@ -290,7 +297,30 @@ function sem(x)
         M = new_M
     end
     var = S / (count - 1)
-    return sqrt(var/count)
+    return sqrt(var / count)
+end
+
+function sem(x, weights::UnitWeights; kwargs...)
+    if length(x) == length(weights)
+        T = promote(eltype(x), eltype(weights))
+        return sem(Iterators.map(T, x); kwargs...)
+    else
+        throw(ArgumentError("data and weights must be the same length"))
+    end
+end
+
+function sem(x, weights::FrequencyWeights)
+    return sqrt(var(x, weights; corrected=true) / weights.sum)
+end
+
+function sem(x, weights::ProbabilityWeights)
+    n = length(x)
+    n == length(weights) || error("data and weights must be the same length")
+    μ = mean(x, weights)
+    var = sum(zip(x, weights)) do (x, w)
+        return (w * (x - μ))^2
+    end
+    return sqrt(var * n / (n - 1)) / weights.sum
 end
```

Comment:
Why not just
```suggestion
sem(x; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))
function sem(x, weights::AbstractWeights; mean=nothing)
    return sqrt(var(x, weights; mean=mean, corrected=true) / sum(weights))
end
```
?

Reply:
@devmotion this would give incorrect results for probability weights. For instance, over-sampling large values and then down-weighting them gives a variance lower than sampling IID values.

Analytic weights are more complicated because the docstring doesn't specify that they _must_ be inverse-variance weights, so it's not clear how they might be generated.

Reply:
The docstring for `AnalyticWeights` says
> These weights may also be referred to as reliability weights, precision weights or inverse variance weights.

So we can definitely assume they must be inverse variance. If that's not clear enough we could improve the docstring. The goal is to have very precise definitions that people can rely on.

Don't you think "reliability weights" mentioned on the Wikipedia page are inverse-variance weights? That sounds plausible, though we would need to check this. Stata supports computing the SEM with analytic weights, but the explanation I found isn't super clear (we want V_srs): https://www.stata.com/support/faqs/statistics/weights-and-summary-statistics/

Reply:
> The docstring for `AnalyticWeights` says
> 
> > These weights may also be referred to as reliability weights, precision weights or inverse variance weights.
> 
> So we can definitely assume they must be inverse variance. If that's not clear enough we could improve the docstring. The goal is to have very precise definitions that people can rely on.
> 
> Don't you think "reliability weights" mentioned on the Wikipedia page are inverse-variance weights? That sounds plausible, though we would need to check this. Stata supports computing the SEM with analytic weights, but the explanation I found isn't super clear (we want V_srs): https://www.stata.com/support/faqs/statistics/weights-and-summary-statistics/

Yes, reliability weights are the same as inverse-variance weights.

I was a bit confused by the docstring, which defines analytic weights as follows:
> Analytic weights describe a non-random relative importance (usually between 0 and 1) for each observation. 

This is a slightly broader definition. General reliability weights might not correspond to the variance per se -- for instance, when weighting observations from several proposal distributions in importance sampling, I might choose to weight them in proportion to `exp(entropy)` instead of the inverse variance. Alternatively, these weights might be taken to refer just to an "Importance" in the sense that predicting some weights poorly might carry a heavier penalty than predicting others. That part of the docstring should probably be edited.

Reply:
Ah yes this is definitely not precise enough. We have the `Weights` type for non-specific cases, so `AnalyticWeights` should be kept for the inverse-variance case (otherwise our estimation of e.g. the variance will likely be wrong).

Reply:
@nalimilan Actually, hold on -- do all of these weights have to be equal to the sample size? i.e. are analytic weights just saying that `x_i` is the average of `w_i = n_i` observations?

Reply:
https://github.com/JuliaStats/StatsBase.jl/pull/757/

Reply:
> #757

Thanks, but this doesn't quite answer my question regarding whether the weights for an `AnalyticWeights` object have to equal the number of observations. The reason being that (according to the Stata source) the variance for analytically-weighted observations depends on the scale of the weights.

Reply:
> @nalimilan Actually, hold on -- do all of these weights have to be equal to the sample size? i.e. are analytic weights just saying that x_i is the average of w_i = n_i observations?

Well, no, because weights are the inverse of the variance, which is proportional to the squared number of observations.

Reply:
> > @nalimilan Actually, hold on -- do all of these weights have to be equal to the sample size? i.e. are analytic weights just saying that x_i is the average of w_i = n_i observations?
> 
> Well, no, because weights are the inverse of the variance, which is proportional to the squared number of observations.

Ahh, ok, so the weights are exactly equal to the inverse of the variance, which must be known ahead of time?

What does `var(x, AnalyticWeights)` do, then? If the inverse-variance for all weights is known, there's nothing to estimate.

Reply:
> Ahh, ok, so the weights are exactly equal to the inverse of the variance, which must be known ahead of time?

Yes, that's my understanding. "Known" or just estimated. (Note that my first reply was wrong, I edited it.)

> What does `var(x, AnalyticWeights)` do, then? If the inverse-variance for all weights is known, there's nothing to estimate.

These are tricky issues. I don't remember the details, but see https://github.com/JuliaStats/StatsBase.jl/pull/250#issuecomment-297317235 and https://github.com/JuliaStats/StatsBase.jl/pull/250#pullrequestreview-35560709. I think the situation for analytic weights is less clear than for other types.

Reply:
> > Ahh, ok, so the weights are exactly equal to the inverse of the variance, which must be known ahead of time?
> 
> Yes, that's my understanding. "Known" or just estimated. (Note that my first reply was wrong, I edited it.)

Estimated weights can get very tricky and require complicated maximum-likelihood estimators and very weird bias corrections; the code we have right now only corrects well in the case that the inverse variance is known ahead of time.

Reply:
> > @nalimilan Actually, hold on -- do all of these weights have to be equal to the sample size? i.e. are analytic weights just saying that x_i is the average of w_i = n_i observations?
> 
> Yes, at least for simple cases that's how it works. Though you can have other situations (e.g. regression coefficients) where variances are estimated using more complex methods. (EDIT: I fixed my previous comment, I was thinking about the standard deviation.)

In that case, it seems that `var` returns incorrect answers -- the results differ quite markedly from those given in the Stata source you linked. Notably, `var` is scale-invariant here (the results are the same regardless of how the weights are scaled).

> Yes, that's my understanding. "Known" or just estimated. (Note that my first reply was wrong, I edited it.)

Known and estimated inverse-precision weights would need different methods and bias corrections. (Actually, I think estimated inverse-precision weights can get quite involved.)

Reply:
I'm thinking it might make sense to do one or more of the following to prevent confusion:
1. Split `AnalyticWeights` into several types of weights. One could be `PrecisionWeights` (where `w_i = 1/σ_i^2` is known ahead of time); another could be `BatchWeights`, where `w_i=n_i` and `x_i` is the average of `n_i` observations. We might also want to split `BatchWeights` depending on whether there's any group-level variation in the mean for each batch of observations.
2. Include sample size as a variable in a set of weights. It seems like part of the confusion here is that the sample size is assumed to be equal to `length(weights)` for analytic weights, which isn't the case when each observation is a minibatch of several observations.

Reply:
I thought about adding more weights types too. I don't think storing the sample size separately would be a good idea, as weights can already carry this information and we wouldn't be able to recompute the sample size when subsetting.

If we add more weights types we need a very clear plan about what these mean and what formulas we can use (for variance and quantile notably). But I'm a bit hesitant as I'm not aware of precedents in other software, so the terminology probably isn't settled and people could get even more confused. If you're interested in investigating this, can you file a separate issue?

Reply:
This reference seems to give everything we need to compute the SEM for analytic weights:
James Kirchner, Weighted averages and their uncertainties
http://seismo.berkeley.edu/~kirchner/Toolkits/Toolkit_12.pdf
(which itself cites Bevington, P. R., Data Reduction and Error Analysis for the Physical Sciences, 336 pp., McGraw-Hill, 1969.)

Reply:
> This reference seems to give everything we need to compute the SEM for analytic weights: James Kirchner, Weighted averages and their uncertainties http://seismo.berkeley.edu/~kirchner/Toolkits/Toolkit_12.pdf (which itself cites Bevington, P. R., Data Reduction and Error Analysis for the Physical Sciences, 336 pp., McGraw-Hill, 1969.)

As mentioned, how to compute the SEM depends on the meaning of the analytic weights. If they are inverse-variance weights with a known variance for instance, the SEM is `sqrt(1/Σw_i)`, which is different from the formula given here.